src/Polysemy/Final.hs

```diff
@@ -0,0 +1,279 @@
+  s <- getInitialStateS
+  pure $ fmap (<$ s) m
+
+------------------------------------------------------------------------------
+-- | Lifts a monadic action into the stateful environment, in terms
+-- of the final monad.
+-- The stateful environment will be the same as the one that the target monad
+-- is initially run in.
+-- Use 'bindS'  if you'd prefer to explicitly manage your stateful environment.
+runS :: Monad m => n a -> Sem (WithStrategy m f n) (m (f a))
+runS = fmap runM . runT
+{-# INLINE runS #-}
+
+------------------------------------------------------------------------------
+-- | Lift a kleisli action into the stateful environment, in terms of the final
+-- monad. You can use 'bindS' to get an effect parameter of the form @a -> m b@
+-- into something that can be used after calling 'runS' on an effect parameter
+-- @m a@.
+bindS :: Monad m => (a -> n b) -> Sem (WithStrategy m f n) (f a -> m (f b))
+bindS = fmap (runM .) . bindT
+{-# INLINE bindS #-}
+
+------------------------------------------------------------------------------
+-- | Internal function to process Strategies in terms of 'withWeaving'.
+runStrategy :: Functor f
+            => f ()
+            -> (forall x. f (n x) -> m (f x))
+            -> (forall x. f x -> Maybe x)
+            -> Sem (WithStrategy m f n) a
+            -> a
+runStrategy s wv ins (Sem m) = runIdentity $ m $ \u -> case extract u of
+  Yo e s' _ ex' _ -> Identity $ ex' $ (<$ s') $ case e of
+    GetInitialState -> s
+    HoistInterpretation na -> sendM . wv . fmap na
+    GetInspector -> Inspector ins
+
+------------------------------------------------------------------------------
+-- Lower a 'Sem' containing only a lifted, final monad into that monad.
+-- The appearance of 'Lift' as the final effect
+-- is to allow the use of operations that rely on a @'LastMember' ('Lift' m)@
+-- constraint.
+runFinal :: Monad m => Sem '[Final m, Lift m] a -> m a
+runFinal (Sem sem) = sem $ \u -> case decomp u of
+  Right (Yo (WithWeaving wav) s wv ex ins) ->
+    ex <$> wav s (runFinal . wv) ins
+  Left g -> case extract g of
+    Yo (Lift m) s _ ex _ -> ex . (<$ s) <$> m
+
+------------------------------------------------------------------------------
+-- Lower a 'Sem' containing two lifted monad into the final monad,
+-- by interpreting one of the monads in terms of the other one.
+--
+-- This allows for the use of operations that rely on a @'LastMember' ('Lift' m)@
+-- constraint, as long as @m@ can be transformed to the final monad;
+-- but be warned, this breaks the implicit contract of @'LastMember' ('Lift' m)@
+-- that @m@ truly _is_ the final monad, so depending on the final monad and operations
+-- used, 'runFinalTrans' may become _unsafe_.
+--
+-- For example, 'runFinalTrans' is unsafe with 'runAsync' if
+-- the final monad is non-deterministic, or a continuation
+-- monad.
+runFinalLift :: Monad m
+              => (forall x. n x -> m x)
+              -> Sem [Final m, Lift m, Lift n] a
```

Comment:
Well, the reasoning was that if you had some `Final` monad that isn't `IO`, but is `MonadIO`, you still couldn't use ops with a `LastMember (Lift IO)` constraint.
In order to use stuff like `runAsync`, the last member of the effect stack must be `Lift IO`, but you can't normally have that if the `Final` monad isn't `IO`.
This is for a very good reason, as `runAsync` typically isn't safe when the end monad isn't directly `IO`, but in some cases, it might still work out fine. `runFinalLift` is for those cases.

However, the only convincing motivational example I had at the time now turns out to be unsafe, so the only things I can point to are simple monad transformers like having `Final (WriterT IO)` but `LastMember (Lift IO)`, in which case using `runFinalLiftIO` together with `runWriterFinal` and `runAsync` _is_ safe... probably.

But that's not very convincing of a use-case. I am willing to drop `runFinalLift` and `runFinalLiftIO`.

The extra `Lift` for `m` is for the sake of consistency, and the ability for interpreters to use `sendM` and not just `withWeaving` and friends on the target monad.